timeboost-sequencer/src/decrypt.rs

```diff
@@ -282,272 +208,318 @@ impl Worker {
+    fn oldest_cached_round(&self) -> RoundNumber {
+        self.dec_shares
+            .keys()
+            .next()
+            .copied()
+            .unwrap_or(RoundNumber::genesis())
+    }
+
+    /// logic to process a `WorkerRequest::Decrypt` request from the decrypter
+    async fn on_decrypt_request(&mut self, round: RoundNumber, incl: InclusionList) -> Result<()> {
+        let dec_shares = self.decrypt(round, incl.clone()).await?;
+        self.net
+            .broadcast(round.u64(), serialize(&dec_shares)?)
+            .await?;
+        self.insert_shares(dec_shares)?;
+        self.incls.insert(round, incl);
+
+        // edge case: when processing the first decrypt request, workers may have received decryption shares
+        // of eariler rounds from other decrypters, but this worker doesn't need those shares, thus pruning them.
+        if self.first_requested_round.is_none() {
+            debug!(node=%self.label, %round, "set first requested round");
+            self.dec_shares.retain(|k, _| k >= &round);
+            self.first_requested_round = Some(round);
+        }
 
-        for (idx, EncryptedItem(kid, data)) in encrypted_items.iter().enumerate() {
-            let ciphertext = deserialize::<Ciphertext>(data)?;
+        Ok(())
+    }
 
-            // establish mappings
-            let cid = ciphertext.nonce();
-            self.cid2ct.insert((round, cid), ciphertext.clone());
-            self.cid2idx.insert(cid, idx);
+    /// garbage collect internally cached state. Different from `self.on_gc_request()` which deals with incoming gc request,
+    /// indicating more than just local gc, but also overlay network-wise gc.
+    fn gc(&mut self, round: RoundNumber) {
+        self.dec_shares.retain(|r, _| *r > round);
+        self.incls.retain(|r, _| *r > round);
+    }
 
+    /// logic to garbage collect a round (and all prior rounds)
+    async fn on_gc_request(&mut self, round: RoundNumber) -> Result<()> {
+        if round.u64() > 0 && !self.dec_shares.is_empty() {
+            self.net.gc(round.u64());
+            self.gc(round);
+        }
+        Ok(())
+    }
+
+    /// scan through the inclusion list and extract the relevant ciphertext from encrypted bundle/tx, preserving the order,
+    /// "relevant" means encrypted under the keyset this worker belongs to.
+    fn extract_ciphertexts(&self, incl: &InclusionList) -> Result<Vec<Ciphertext>> {
+        let kid = self.committee.id();
+        let cts = incl
+            .priority_bundles()
+            .iter()
+            .filter(|pb| pb.bundle().kid() == Some(kid))
+            .map(|pb| pb.bundle().data())
+            .chain(
+                incl.regular_bundles()
+                    .iter()
+                    .filter(|b| b.kid() == Some(kid))
+                    .map(|b| b.data()),
+            )
+            .map(|bytes| deserialize::<Ciphertext>(bytes))
+            .collect::<Result<Vec<_>>>()?;
```

Comment:
Would this not mean that a single decryption failure of one bundle spoils the whole inclusion list? Perhaps we should drop those bundles what failed to deserialise successfully, but continue with the remaining ones.

Reply:
that's a good point, then I guess we just treat them as garbage and still mark them as decrypted? (by resetting the `kid: None`)

Reply:
done in [b2b39b3](https://github.com/EspressoSystems/timeboost/pull/346/commits/b2b39b3039dcf9dc6e9fe6ade616d6f9146c55c4)